src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -169,21 +197,73 @@ private void RegisterNotification(Action action)
 
             private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
             {
-                OnDiagnosticsUpdated(e);
+                lock (_taskGate)
+                {
+                    // Chain the events so we process them serially.  This also ensures
+                    // that we don't process events while still getting our initial set
+                    // of diagnostics.
+                    var asyncToken = _owner._listener.BeginAsyncOperation(GetType() + ".OnDiagnosticsUpdated");
+                    _taskChain = _taskChain.ContinueWith(
+                        _ => OnDiagnosticsUpdatedOnBackground(e), TaskScheduler.Default);
+                    _taskChain.CompletesAsyncOperation(asyncToken);
+                }
+            }
+
+            private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
+            {
+                this.AssertIsBackground();
+                RegisterNotification(() => OnDiagnosticsUpdatedOnForeground(e));
```

Comment:
We dont' want to ContinueWith(+fgScheduler).  That would mean we'd have to know how RegisterNotification works (for example, if it has any delays built into it).  For example, we could have this sequence of calls:

Task1 does:
```
GetInitialDiagnosticsOnBackground.
RegisterNotification(() => OnDiagnosticsUpdatedOnForeground())
```

Task2 does:
```
ContinueWith(fgScheduler)
```
That ContinueWith may execute before whatever RegisterNotification eventually executes.

Indeed, RegisterNotification has a default delay of 15ms built into it.  So this could cause a race.

In essence, RegisterNotification acts as a 'SyncContext'.  And if not all codepaths are going through it, we're going to race conditions.